sdks/java/io/solr/src/main/java/org/apache/beam/sdk/io/solr/SolrIO.java

```diff
@@ -0,0 +1,641 @@
+import org.apache.solr.client.solrj.impl.HttpSolrClient;
+import org.apache.solr.client.solrj.request.CoreAdminRequest;
+import org.apache.solr.client.solrj.response.CoreAdminResponse;
+import org.apache.solr.client.solrj.response.QueryResponse;
+import org.apache.solr.common.SolrDocument;
+import org.apache.solr.common.SolrInputDocument;
+import org.apache.solr.common.cloud.ClusterState;
+import org.apache.solr.common.cloud.DocCollection;
+import org.apache.solr.common.cloud.Replica;
+import org.apache.solr.common.cloud.Slice;
+import org.apache.solr.common.params.CoreAdminParams;
+import org.apache.solr.common.params.CursorMarkParams;
+import org.apache.solr.common.util.JavaBinCodec;
+import org.apache.solr.common.util.NamedList;
+
+/**
+ * Transforms for reading and writing data from/to Solr.
+ *
+ * <h3>Reading from Solr</h3>
+ *
+ * <p>{@link SolrIO#read SolrIO.read()} returns a bounded
+ * {@link PCollection PCollection&lt;SolrDocument&gt;} representing Solr documents.
+ *
+ * <p>To configure the {@link SolrIO#read}, you have to provide a connection configuration
+ * containing the Zookeeper address of the Solr cluster. The following example
+ * illustrates options for configuring the source:
+ *
+ * <pre>{@code
+ *
+ * pipeline.apply(SolrIO.read().withConnectionConfiguration(
+ *    SolrIO.ConnectionConfiguration.create("127.0.0.1:9983", "my-collection")
+ * )
+ *
+ * }</pre>
+ *
+ * <p>You can specify a query on the {@code read()} using {@code withQuery()}.
+ *
+ * <h3>Writing to Solr</h3>
+ *
+ * <p>To write documents to Solr, use
+ * {@link SolrIO#write SolrIO.write()}, which writes Solr documents from a
+ * {@link PCollection PCollection&lt;SolrInputDocument&gt;} (which can be bounded or unbounded).
+ *
+ * <p>To configure {@link SolrIO#write SolrIO.write()}, similar to the read, you
+ * have to provide a connection configuration. For instance:
+ *
+ * <pre>{@code
+ *
+ *  pipeline
+ *    .apply(...)
+ *    .apply(SolrIO.write().withConnectionConfiguration(
+ *       SolrIO.ConnectionConfiguration.create("127.0.0.1:9983", "my-collection")
+ *    )
+ *
+ * }</pre>
+ *
+ * <p>Optionally, you can provide {@code withBatchSize()}
+ * to specify the size of the write batch in number of documents.
+ */
+@Experimental
+public class SolrIO {
+
+  public static Read read() {
+    return new AutoValue_SolrIO_Read.Builder().setBatchSize(100L).build();
+  }
+
+  public static Write write() {
+    return new AutoValue_SolrIO_Write.Builder().setMaxBatchSize(1000L).build();
+  }
+
+  private SolrIO() {
+  }
+
+  /** A POJO describing a connection configuration to Solr. */
+  @AutoValue public abstract static class ConnectionConfiguration implements Serializable {
+
+    //TODO add user name password
+    abstract String getZkHost();
+
+    abstract String getCollection();
+
+    abstract Builder builder();
+
+    @AutoValue.Builder abstract static class Builder {
+
+      abstract Builder setZkHost(String zkHost);
```

Comment:
Yes, feel free to create some JIRA issues for this and other future work. I didn't know that 7.0 fixes this, in any case it is more important to cover the parallel case so better to have the CloudSolr version first.